python/pyiceberg/io/pyarrow.py

```diff
@@ -546,11 +568,45 @@ def project_table(
         id for id in projected_schema.field_ids if not isinstance(projected_schema.find_type(id), (MapType, ListType))
     }.union(extract_field_ids(bound_row_filter))
 
+    tasks_data_files: List[FileScanTask] = []
+    tasks_positional_deletes: List[FileScanTask] = []
+    for task in tasks:
```

Comment:
Scan planning should use an index of delete files to find delete files for each file task. Then the file task has a data file to read and deletes to apply. That process should handle sequence number filtering so that we don't apply deletes to files that have newer sequence numbers.

At this point, all we should need to do is to read the delete files, filter the deletes by the data file path, then merge the resulting delete vectors together. Then do the mask-based filter that you have above.

Reply:
Let's follow up and talk about how we want to do this in Python.

Reply:
> Scan planning should use an index of delete files to find delete files for each file task. Then the file task has a data file to read and deletes to apply. That process should handle sequence number filtering so we don't apply deletes to files with newer sequence numbers.

I agree, but this isn't relevant for positional deletes.

Looking at the partition pruning, done here: https://github.com/apache/iceberg/blob/2e45760c8d2b62cc46584cbff28bcb2df422d7ff/core/src/main/java/org/apache/iceberg/DeleteFileIndex.java#L529-L556

I checked and we already prune on a partition level for Python, and I'm not sure why there is a different path in Java.

Reply:
https://www.loom.com/share/2afdfd79d49a46ff9c3d8db17ba87cfa